pkg/virtual/framework/transforming/types.go

```diff
@@ -0,0 +1,305 @@
+
+	BeforeUpdate func(ctx context.Context, obj *unstructured.Unstructured, options metav1.UpdateOptions, subresources ...string) (context.Context, *unstructured.Unstructured, metav1.UpdateOptions, []string, error)
+	AfterUpdate  func(ctx context.Context, obj *unstructured.Unstructured, options metav1.UpdateOptions, subresources []string, result *unstructured.Unstructured) (*unstructured.Unstructured, error)
+
+	BeforeDelete           func(ctx context.Context, name string, options metav1.DeleteOptions, subresources ...string) (context.Context, string, metav1.DeleteOptions, []string, error)
+	BeforeDeleteCollection func(ctx context.Context, options metav1.DeleteOptions, listOptions metav1.ListOptions) (context.Context, metav1.DeleteOptions, metav1.ListOptions, error)
+
+	BeforeGet func(ctx context.Context, name string, options metav1.GetOptions, subresources ...string) (context.Context, string, metav1.GetOptions, []string, error)
+	AfterGet  func(ctx context.Context, name string, options metav1.GetOptions, subresources []string, result *unstructured.Unstructured) (*unstructured.Unstructured, error)
+
+	BeforeList func(ctx context.Context, opts metav1.ListOptions) (context.Context, metav1.ListOptions, error)
+	AfterList  func(ctx context.Context, opts metav1.ListOptions, result *unstructured.UnstructuredList) (*unstructured.UnstructuredList, error)
+
+	BeforeWatch func(ctx context.Context, opts metav1.ListOptions) (context.Context, metav1.ListOptions, error)
+	AfterWatch  func(ctx context.Context, opts metav1.ListOptions, result watch.Interface) (watch.Interface, error)
+}
+
+type TransformingClient struct {
+	Transformations Transformers
+	Client          dynamic.ResourceInterface
+	Resource        schema.GroupVersionResource
+	Namespace       string
+}
+
+func (tc *TransformingClient) objectCallMessage(transformerName, action string, obj *unstructured.Unstructured, subresources ...string) string {
+	name := "nil"
+	if obj != nil {
+		name = obj.GetName()
+	}
+	return tc.namedCallMessage(transformerName, action, name, subresources...)
+}
+
+func (tc *TransformingClient) namedCallMessage(transformerName, action string, name string, subresources ...string) string {
+	return fmt.Sprintf("%s(%s): %s/%s(%s) - %v ", action, transformerName, tc.Namespace, name, tc.Resource, subresources)
+}
+
+func (tc *TransformingClient) logObjectCall(transformerName, action string, obj *unstructured.Unstructured, subresources ...string) {
+	klog.Info(tc.objectCallMessage(transformerName, action, obj, subresources...))
+}
+
+func (tc *TransformingClient) logNamedCall(transformerName, action string, name string, subresources ...string) {
+	klog.Info(tc.namedCallMessage(transformerName, action, name, subresources...))
+}
+
+func (tc *TransformingClient) logCallError(transformerName, action string, err error) {
+	klog.Errorf("Transformation Error on %s(%s): %v", action, transformerName, err)
+}
```

Comment:
is this all debug code?

Reply:
This is common log-related code to have consistent and complete logging of transformations